File: backend/src/test/kotlin/org/globe42/web/persons/WeddingEventControllerMvcTest.java
package org.globe42.web.persons;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.time.LocalDate;
import java.util.Optional;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.globe42.dao.PersonDao;
import org.globe42.domain.Person;
import org.globe42.domain.WeddingEvent;
import org.globe42.domain.WeddingEventType;
import org.globe42.test.GlobeMvcTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC tests for {@link WeddingEventController}
 * @author JB Nizet
 */
@GlobeMvcTest(WeddingEventController.class)
public class WeddingEventControllerMvcTest {
    @MockBean
    private PersonDao mockPersonDao;

    @Autowired
    private MockMvc mvc;

    @Autowired
    private ObjectMapper objectMapper;

    private Person person;
    private WeddingEvent firstWedding;

    @BeforeEach
    public void prepare() {
        person = new Person(42L);
        firstWedding = new WeddingEvent(34L);
        firstWedding.setDate(LocalDate.of(2000, 2, 28));
        firstWedding.setType(WeddingEventType.WEDDING);
        person.addWeddingEvent(firstWedding);
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));
    }

    @Test
    public void shouldList() throws Exception {
        mvc.perform(get("/api/persons/{personId}/wedding-events", person.getId()))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(34))
           .andExpect(jsonPath("$[0].date").value("2000-02-28"))
           .andExpect(jsonPath("$[0].type").value(WeddingEventType.WEDDING.name()));
    }

    @Test
    public void shouldCreate() throws Exception {
        WeddingEventCommandDTO command = new WeddingEventCommandDTO(LocalDate.of(2002, 3, 28),
                                                                    WeddingEventType.DIVORCE);
        mvc.perform(post("/api/persons/{personId}/wedding-events", person.getId())
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.type").value(WeddingEventType.DIVORCE.name()));
    }

    @Test
    public void shouldDelete() throws Exception {
        mvc.perform(delete("/api/persons/{personId}/wedding-events/{eventId}", person.getId(), firstWedding.getId()))
           .andExpect(status().isNoContent());
    }
}

File: backend/src/test/kotlin/org/globe42/web/persons/WeddingEventControllerMvcTest.kt
package org.globe42.web.persons

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.globe42.dao.PersonDao
import org.globe42.domain.*
import org.globe42.test.GlobeMvcTest
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.delete
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import java.time.LocalDate
import java.util.*

/**
 * MVC tests for [WeddingEventController]
 * @author JB Nizet
 */
@GlobeMvcTest(WeddingEventController::class)
class WeddingEventControllerMvcTest(
    @Autowired private val mvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {
    @MockkBean
    private lateinit var mockPersonDao: PersonDao

    private lateinit var person: Person
    private lateinit var firstWedding: WeddingEvent

    @BeforeEach
    fun prepare() {
        person = Person(42L, "John", "Doe", Gender.MALE)
        firstWedding = WeddingEvent(34L)
        firstWedding.date = LocalDate.of(2000, 2, 28)
        firstWedding.type = WeddingEventType.WEDDING
        firstWedding.location = Location.ABROAD
        person.addWeddingEvent(firstWedding)
        every { mockPersonDao.findById(person.id!!)} returns Optional.of(person)
    }

    @Test
    fun `should list`() {
        mvc.get("/api/persons/{personId}/wedding-events", person.id).andExpect {
            status { isOk }
            jsonValue("$[0].id", 34)
            jsonValue("$[0].date", "2000-02-28")
            jsonValue("$[0].type", WeddingEventType.WEDDING.name)
            jsonValue("$[0].location", Location.ABROAD.name)
        }
    }

    @Test
    fun `should create`() {
        val date = LocalDate.of(2002, 3, 28)
        val command = WeddingEventCommandDTO(
            date,
            WeddingEventType.DIVORCE,
            Location.FRANCE
        )
        every { mockPersonDao.flush() } answers {
            person.getWeddingEvents().find { it.date == date }?.let { it.id = 876 }
            Unit
        }
        mvc.post("/api/persons/{personId}/wedding-events", person.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isCreated }
            jsonValue("$.type", WeddingEventType.DIVORCE.name)
        }
    }

    @Test
    fun `should delete`() {
        mvc.delete("/api/persons/{personId}/wedding-events/{eventId}", person.id, firstWedding.id).andExpect {
            status { isNoContent }
        }
    }
}
